File: 10. Regular Expression Matching Hard/solution.java
class solution {
    public static boolean isMatch(String s, String p) {
        int m = s.length();
        int n = p.length();
        boolean[][] memo = new boolean[m + 1][n + 1];
        memo[m][n] = true;

        for (int i = m; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                boolean firstMatch = i < m && (p.charAt(j) == s.charAt(i) || p.charAt(j) == '.');
                if (j + 1 < n && p.charAt(j + 1) == '*') {
                    memo[i][j] = memo[i][j + 2] || (firstMatch && memo[i + 1][j]);
                } else {
                    memo[i][j] = firstMatch && memo[i + 1][j + 1];
                }
            }
        }

        return memo[0][0];
    }

    public static void main(String[] args) {
        System.out.println(isMatch("aa", "a"));
        System.out.println(isMatch("aa", "a*"));
        System.out.println(isMatch("ab", ".*"));
    }
}

File: 10. Regular Expression Matching Hard/solution.ts
function isMatch(s: string, p: string): boolean {
    const memo: {[key: string]: boolean} = {};

    function dp(i: number, j: number): boolean {
        const key = `${i},${j}`;
        if (memo[key] !== undefined) {
            return memo[key];
        }

        if (j === p.length) {
            const ans = i === s.length;
            memo[key] = ans;
            return ans;
        } else {
            const firstMatch = i < s.length && (p[j] === s[i] || p[j] === '.');

            if (j + 1 < p.length && p[j + 1] === '*') {
                const ans = dp(i, j + 2) || (firstMatch && dp(i + 1, j));
                memo[key] = ans;
                return ans;
            } else {
                const ans = firstMatch && dp(i + 1, j + 1);
                memo[key] = ans;
                return ans;
            }
        }
    }

    return dp(0, 0);
}

console.log(isMatch('aa', 'a'));
console.log(isMatch('aa', 'a*'));
console.log(isMatch('ab', '.*'));